feat: add how-to guide on chiselled docker image (#11)
canonical/chisel-docs: docs/how-to/create-chiselled-docker-image.md

> # Create a chiselled Docker image

Author: Done.

> Since Chisel can produce a minimal root file system, using it in Docker images

Author: Changed.

has lots of value. This guide shows how to create a chiselled Docker image.

> A _chiselled_ image is one which only contains files installed by the slices and

Reviewer: I'd remove this. it's unnecessarily conveying double meaning to the word "chiseled" vs the fact that is must be baseless (which is not true - but rather more like an intuitive pattern to follow)

Author: Removed.

> A _chiselled_ image is one which only contains files installed by the slices and

Author: Ya thanks for that. Changed it.

does not contain any base images such as the ubuntu image.

## Design the image

> It's a good idea to think first about what our chiselled image should be able to

Reviewer: Too much fluff IMO. Just be short and straight to the point.

do. Let's consider a Python 3 image, which contains a Python 3 interpreter and
is able to run Python 3 scripts and commands.

Goals:

- The [image entrypoint] must be `python3`.

> - The image must be chiselled i.e. there must not be anything else.

Reviewer: see the comment above about the double meaning of the word "chiseled". It's also redundant with the other bullet below about being  "minimal"

Author: Removed.

- The image must be able to run Python 3 scripts.
- The image must be minimal.

To build this image, we will use [multi-stage docker builds].

In earlier stage(s),

1. We will install Chisel and other dependencies that we may
need.
2. We will install the slices using Chisel and prepare a root file system.

> Finally, we will copy the root file system to the `/` directory in the final

Reviewer: ```suggestion
Finally, we will copy the staging area's root file system to the `/` directory in the final
```

> stage. The final stage base image will be [`scratch`], i.e. nothing. Thus, the

Reviewer: ```suggestion
stage. The final stage's base image will be [`scratch`]. Thus, the
```

image only contains the root file system installed by Chisel and nothing else.

> ## Install Chisel and dependencies

Reviewer: ```suggestion
## Write the Dockerfile

Create a Dockerfile and open it with your favorite text editor.

### Install Chisel and dependencies
```

To make building easier and configurable, let's first define some variables that
we can re-use.

```docker
ARG UBUNTU_RELEASE=24.04

> ARG USER=app UID=101 GROUP=app GID=101

Reviewer: don't use these low UIDs

Author: Removed them altogether.

Reviewer: Well, I don't think that's better, cause then you're building a root container, which is a bad practice. Especially now that you are installing the base-passwd.

Author: Alright, I resorted back to using the `ubuntu` user and removed the `base-passwd` slices.

ARG CHISEL_VERSION=1.1.0
```

Let's now initialize a new build stage where we install Chisel and dependencies

> and prepare the root file system.

Reviewer: ```suggestion
and prepare the final chiseled root file system.
```

```docker
FROM ubuntu:$UBUNTU_RELEASE AS builder
ARG USER UID GROUP GID TARGETARCH UBUNTU_RELEASE CHISEL_VERSION
SHELL ["/bin/bash", "-oeux", "pipefail", "-c"]
```

> We will need the `ca-certificates` package to fetch files. Let's install that.

Reviewer: ```suggestion
Chisel needs the `ca-certificates` package in order to fetch files from the archives. Let's install that.
```

```docker
RUN apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y ca-certificates \
    && apt-get clean -y \

>     && rm -rf /var/lib/apt/lists/*

Reviewer: I'd do some more research about this. I know it's common practice, but I always frowned upon this practice within a multi stage build, cause it only slows down the build, and it's irrelevant (since this is just a build stage). 

If seen as a best practice, then keep it

Author: It's a best practice, indicated in the docker docs: https://docs.docker.com/build/building/best-practices/#apt-get

Reviewer: right, I know, but that page isn't specific to multi-stage. One should definitely do this in the target stages. But in intermediary stages, what's the drawback? larger cache maybe...but additional instructions polluting the dockerfile and slowing down the build...anyway, it's a nitpick, forget about it

```

We can now install Chisel. Make sure to install the latest version. Visit the
{{latest_release_page}} to determine the latest version and override

> `CHISEL_VERSION` appropriately.

Reviewer: this comment should probably be above where you define the ARG

Author: Relocated.

```docker
ADD "https://github.com/canonical/chisel/releases/download/v${CHISEL_VERSION}/chisel_v${CHISEL_VERSION}_linux_${TARGETARCH}.tar.gz" \
    chisel.tar.gz
RUN tar -xvf chisel.tar.gz -C /usr/bin/
```

> ## Prepare chiselled root file system

Reviewer: ```suggestion
## Prepare the chiseled root file system
```

Now that we have Chisel installed in the `builder` stage, we can install the
appropriate slices and prepare the root file system.

The [`python3_standard` slice] provides `python3`. We will install this slice,
along with a few others:

- `base-files_base` for file structure.
- `base-files_release-info` for release info.
- `base-files_chisel` for the chisel manifest.
- `ca-certificates_data` for network support.

```docker

> RUN mkdir /rootfs \

Reviewer: can we call it `/staging-rootfs` instead?

    && chisel cut --release "ubuntu-$UBUNTU_RELEASE" --root /rootfs \
        base-files_base \
        base-files_release-info \
        base-files_chisel \
        ca-certificates_data \
        python3_standard
```

We will also install two users in the root file system - `USER` and `root`.

```docker
RUN install -d -m 0755 -o "$UID" -g "$GID" "/rootfs/home/$USER" \
    && echo -e "root:x:0:\n$GROUP:x:$GID:" >/rootfs/etc/group \
    && echo -e "root:x:0:0:root:/root:/noshell\n$USER:x:$UID:$GID::/home/$USER:/noshell" >/rootfs/etc/passwd

> ```

Reviewer: why not use the `base-passwd` slices?

Reviewer: i'd also be ok with jsut copying the /etc/passwd from the ubuntu builder stage, since it already has the `ubuntu:1000` user that we can use

Author: Used `base-passwd`, thanks!

> ## Copy the root file system to the final image

Reviewer: ```suggestion
### Copy the chiseled root file system to the final image
```

Finally, we will initialize a new stage where we will copy the prepared root
file system (`/rootfs`) from the previous stage.

```docker
FROM scratch
ARG USER UID GROUP GID

COPY --from=builder /rootfs /
# Workaround for https://github.com/moby/moby/issues/38710
COPY --from=builder --chown="$UID:$GID" "/rootfs/home/$USER" "/home/$USER"
```

> ## Set the entrypoint

Reviewer: ```suggestion
### Set the entrypoint
```

> All the remains is to set the entrypoint to `python3`.

Reviewer: You're also missing the nonroot `USER`

Author: Removed users altogther.

Author: Fixed.

```docker
ENTRYPOINT ["/usr/bin/python3"]
```

## Build and test the image

Let's copy all the code-blocks above and save it to `Dockerfile`. Run the

> following command to build the image.

Reviewer: it would be nice to see the entire Dockerfile
